x-pack/filebeat/tests/integration/managerV2_test.go

```diff
@@ -344,12 +344,162 @@ func TestFailedOutputReportsUnhealthy(t *testing.T) {
+		Type:           proto.UnitType_INPUT,
+		ConfigStateIdx: 1,
+		State:          proto.State_STARTING,
+		LogLevel:       logLevel,
+		Config: &proto.UnitExpectedConfig{
+			Id:   "filestream-input",
+			Type: "filestream",
+			Name: "filestream-input-starting",
+			Streams: []*proto.Stream{
+				{
+					Id: "filestream-input-id",
+					Source: integration.RequireNewStruct(t, map[string]interface{}{
+						"id":      "filestream-stream-input-id",
+						"enabled": true,
+						"type":    "filestream",
+						"paths":   logFilePath,
+					}),
+				},
+			},
+		},
+	}
+
+	healthyOutput := proto.UnitExpected{
+		Id:             "output-unit",
+		Type:           proto.UnitType_OUTPUT,
+		ConfigStateIdx: 1,
+		State:          proto.State_HEALTHY,
+		LogLevel:       logLevel,
+		Config: &proto.UnitExpectedConfig{
+			Id:   "default",
+			Type: "elasticsearch",
+			Name: "elasticsearch",
+			Source: integration.RequireNewStruct(t,
+				map[string]interface{}{
+					"type":     "elasticsearch",
+					"hosts":    []interface{}{"http://localhost:9200"},
+					"username": "admin",
+					"password": "testing",
+					"protocol": "http",
+					"enabled":  true,
+				}),
+		},
 	}
-	return str
+
+	brokenOutput := proto.UnitExpected{
+		Id:             "output-unit-borken",
+		Type:           proto.UnitType_OUTPUT,
+		ConfigStateIdx: 1,
+		State:          proto.State_FAILED,
+		LogLevel:       logLevel,
+		Config: &proto.UnitExpectedConfig{
+			Id:   "default",
+			Type: "logstash",
+			Name: "logstash",
+			Source: integration.RequireNewStruct(t,
+				map[string]interface{}{
+					"type":    "logstash",
+					"invalid": "configuration",
+				}),
+		},
+	}
+
+	// Those are the 'states' Filebeat will go through.
+	// After each state is reached the mockServer will
+	// send the next.
+	protoUnits := [][]*proto.UnitExpected{
+		{
+			&healthyOutput,
+			&filestreamInputHealthy,
+		},
+		{
+			&brokenOutput,
+			&filestreamInputStarting,
+		},
+		{
+			&healthyOutput,
+			&filestreamInputHealthy,
+		},
+		{}, // An empty one makes the Beat exit
+	}
+
+	// We use `testDone` to notify `require.Eventually` at the end of the test
+	// that the desired state has been reached.
+	testDone := atomic.MakeBool(false)
+	// The test is successful when we reach the last element of `protoUnits`
+	onObserved := func(observed *proto.CheckinObserved, protoUnitsIdx int) {
+		if protoUnitsIdx == len(protoUnits)-1 {
+			testDone.Store(true)
+		}
+	}
+
+	server := integration.NewMockServer(
+		protoUnits,
+		[]uint64{0, 0, 0, 0},
+		[]*proto.Features{nil, nil, nil, nil},
+		onObserved,
+		100*time.Millisecond,
+	)
+	require.NoError(t, server.Start(), "could not start the mock Elastic-Agent server")
+	defer server.Stop()
+
+	filebeat.RestartOnBeatOnExit = true
+	filebeat.Start(
+		"-E", fmt.Sprintf(`management.insecure_grpc_url_for_testing="localhost:%d"`, server.Port),
+		"-E", "management.enabled=true",
+		"-E", "management.restart_on_output_change=true",
+	)
+
+	require.Eventually(
```

Comment:
Could this and the atomic bool just be replaced with a done channel that is closed when the condition is hit?

Then rather than having to poll an atomic bool the test would complete as soon as the condition was true?

This eventually would become a select statement with a timeout:

```go
select {
case <-done:
case <-time.After(60*time.Second)
   require.Fail(t, "Output did not recover from a invalid configuration")
}
```

Reply:
It indeed can! Thanks a lot :D, I'll implement it.